File: semaphore.py
import threading
import time
import random

# ================= PRODUCER–CONSUMER PROBLEM =================
class SharedBuffer:
    def __init__(self):
        self.item = None  # Shared data between producer and consumer
        self.mutex = threading.Semaphore(1)  # Ensures mutual exclusion
        self.empty = threading.Semaphore(1)  # Buffer starts empty
        self.full = threading.Semaphore(0)   # Buffer starts not full

    def produce(self, value):
        self.empty.acquire()  # Wait if buffer is full
        self.mutex.acquire()  # Lock the critical section
        self.item = value
        print(f"Producer produced: {self.item}")
        self.mutex.release()  # Unlock critical section
        self.full.release()   # Signal that buffer has data

    def consume(self):
        self.full.acquire()   # Wait if buffer is empty
        self.mutex.acquire()  # Lock the critical section
        print(f"Consumer consumed: {self.item}")
        self.mutex.release()  # Unlock critical section
        self.empty.release()  # Signal that buffer is empty


class Producer(threading.Thread):
    def __init__(self, buffer):
        super().__init__()
        self.buffer = buffer

    def run(self):
        for i in range(1, 6):
            self.buffer.produce(i)
            time.sleep(1)  # Simulate production time


class Consumer(threading.Thread):
    def __init__(self, buffer):
        super().__init__()
        self.buffer = buffer

    def run(self):
        for _ in range(1, 6):
            self.buffer.consume()
            time.sleep(1.5)  # Simulate consumption time


# ================= DINING PHILOSOPHERS PROBLEM =================
class DiningPhilosophers:
    forks = [threading.Semaphore(1) for _ in range(5)]  # One semaphore per fork
    mutex = threading.Semaphore(1)  # To avoid deadlock

    class Philosopher(threading.Thread):
        def __init__(self, id):
            super().__init__()
            self.id = id

        def run(self):
            for _ in range(3):  # Each philosopher eats 3 times
                print(f"Philosopher {self.id} is thinking.")
                time.sleep(random.uniform(0.5, 1.5))

                # Acquire mutex before picking forks to avoid deadlock
                DiningPhilosophers.mutex.acquire()
                DiningPhilosophers.forks[self.id].acquire()          # Pick left fork
                DiningPhilosophers.forks[(self.id + 1) % 5].acquire()  # Pick right fork
                DiningPhilosophers.mutex.release()

                print(f"Philosopher {self.id} is eating.")
                time.sleep(random.uniform(0.5, 1.5))

                # Release both forks
                DiningPhilosophers.forks[self.id].release()
                DiningPhilosophers.forks[(self.id + 1) % 5].release()

                print(f"Philosopher {self.id} finished eating.")

    @staticmethod
    def start_dining():
        philosophers = [DiningPhilosophers.Philosopher(i) for i in range(5)]
        for p in philosophers:
            p.start()
        return philosophers


# ================= MAIN METHOD =================
if __name__ == "__main__":
    # --- Producer–Consumer Problem Execution ---
    print("=== PRODUCER–CONSUMER PROBLEM ===")
    buffer = SharedBuffer()
    producer = Producer(buffer)
    consumer = Consumer(buffer)

    producer.start()
    consumer.start()

    producer.join()
    consumer.join()

    # --- Dining Philosophers Problem Execution ---
    print("\n=== DINING PHILOSOPHERS PROBLEM ===")
    philosophers = DiningPhilosophers.start_dining()

    # Wait for all philosophers to finish
    for p in philosophers:
        p.join()

File: Semaphore.java

import java.util.concurrent.Semaphore;
public class Semaphore1 {
    // ================= PRODUCER–CONSUMER PROBLEM =================
    static class SharedBuffer {
        int item;  // Shared data between producer and consumer

        // Semaphores for synchronization
        Semaphore mutex = new Semaphore(1);   // Ensures mutual exclusion
        Semaphore empty = new Semaphore(1);   // Buffer starts empty
        Semaphore full = new Semaphore(0);    // Buffer starts not full

        // Producer method
        public void produce(int value) throws InterruptedException {
            empty.acquire();   // Wait if buffer is full
            mutex.acquire();   // Lock the critical section
            item = value;
            System.out.println("Producer produced: " + item);
            mutex.release();   // Unlock critical section
            full.release();    // Signal that buffer has data
        }

        // Consumer method
        public void consume() throws InterruptedException {
            full.acquire();    // Wait if buffer is empty
            mutex.acquire();   // Lock the critical section

            System.out.println("Consumer consumed: " + item);

            mutex.release();   // Unlock critical section
            empty.release();   // Signal that buffer is empty
        }
    }

    // Producer thread
    static class Producer extends Thread {
        SharedBuffer buffer;
        Producer(SharedBuffer b) { buffer = b; }

        public void run() {
            for (int i = 1; i <= 5; i++) {
                try {
                    buffer.produce(i);
                    Thread.sleep(1000); // Simulate production time
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    // Consumer thread
    static class Consumer extends Thread {
        SharedBuffer buffer;
        Consumer(SharedBuffer b) { buffer = b; }

        public void run() {
            for (int i = 1; i <= 5; i++) {
                try {
                    buffer.consume();
                    Thread.sleep(1500); // Simulate consumption time
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    // ================= DINING PHILOSOPHERS PROBLEM =================
    static class DiningPhilosophers {
        static Semaphore[] forks = new Semaphore[5]; // One semaphore per fork
        static Semaphore mutex = new Semaphore(1);   // To avoid deadlock

        // Initialize forks
        static {
            for (int i = 0; i < 5; i++) {
                forks[i] = new Semaphore(1);
            }
        }

        // Philosopher thread
        static class Philosopher extends Thread {
            int id;
            Philosopher(int id) { this.id = id; }

            public void run() {
                try {
                    for (int i = 0; i < 3; i++) {
                        System.out.println("Philosopher " + id + " is thinking.");
                        Thread.sleep((int)(Math.random() * 1000));

                        // Acquire mutex before picking forks (avoid deadlock)
                        mutex.acquire();
                        forks[id].acquire();                // Pick left fork
                        forks[(id + 1) % 5].acquire();      // Pick right fork
                        mutex.release();

                        System.out.println("Philosopher " + id + " is eating.");
                        Thread.sleep((int)(Math.random() * 1000));

                        // Release both forks after eating
                        forks[id].release();
                        forks[(id + 1) % 5].release();

                        System.out.println("Philosopher " + id + " finished eating.");
                    }
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }

        // Start all philosopher threads
        public static Philosopher[] startDining() {
            Philosopher[] philosophers = new Philosopher[5];
            for (int i = 0; i < 5; i++) {
                philosophers[i] = new Philosopher(i);
                philosophers[i].start();
            }
            return philosophers;
        }
    }

    // ================= MAIN METHOD =================
    public static void main(String[] args) {
        // --- Producer–Consumer Problem Execution ---
        System.out.println("=== PRODUCER–CONSUMER PROBLEM ===");
        SharedBuffer buffer = new SharedBuffer();
        Producer producer = new Producer(buffer);
        Consumer consumer = new Consumer(buffer);


        producer.start();
        consumer.start();

        try {
            producer.join();
            consumer.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        // --- Dining Philosophers Problem Execution ---
        System.out.println("\n=== DINING PHILOSOPHERS PROBLEM ===");
        DiningPhilosophers.Philosopher[] philosophers = DiningPhilosophers.startDining();

        // Wait for all philosophers to finish
        for (DiningPhilosophers.Philosopher p : philosophers) {
            try {
                p.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}
